File: packages/cli/src/cli/providers.ts
#!/usr/bin/env bun
import * as p from "@clack/prompts";
import pc from "picocolors";
import { getAllPresets, getBuiltInPresets } from "../providers/presets.js";
import {
  upsertUserProvider,
  deleteUserProvider,
  getUserProvider,
  loadUserProviders,
} from "../providers/user-providers.js";
import type { ProviderPreset } from "../types.js";
import { showError, showSuccess } from "../utils/ui.js";
import { MARKERS } from "../constants/formatting.js";

/**
 * List all available providers
 */
export async function listProviders(): Promise<void> {
  console.log();
  p.intro(pc.cyan("Available Providers"));

  const allPresets = await getAllPresets();
  const userProviders = await loadUserProviders();
  const userProviderIds = new Set(userProviders.map(p => p.id));
  const builtInProviderIds = new Set(getBuiltInPresets().map(p => p.id));

  if (allPresets.length === 0) {
    console.log(pc.yellow("No providers configured"));
    console.log();
    return;
  }

  console.log();
  console.log(pc.bold("Built-in Providers:"));
  allPresets
    .filter(p => builtInProviderIds.has(p.id) && !userProviderIds.has(p.id))
    .forEach(preset => {
      console.log(
        `  ${pc.cyan(preset.id.padEnd(15))} ${pc.dim("|")} ${preset.displayName.padEnd(30)} ${pc.dim("|")} ${pc.yellow(preset.baseURL || "N/A")}`
      );
    });

  const userOnlyProviders = allPresets.filter(p => userProviderIds.has(p.id));
  if (userOnlyProviders.length > 0) {
    console.log();
    console.log(pc.bold("User-defined Providers:"));
    userOnlyProviders.forEach(preset => {
      const isOverride = builtInProviderIds.has(preset.id);
      const marker = isOverride ? pc.yellow(" (override)") : "";
      console.log(
        `  ${pc.green(preset.id.padEnd(15))} ${pc.dim("|")} ${preset.displayName.padEnd(30)} ${pc.dim("|")} ${pc.yellow(preset.baseURL || "N/A")}${marker}`
      );
    });
  }

  console.log();
  console.log(pc.dim(`Total: ${allPresets.length} providers`));
  console.log();
}

/**
 * Add or update a provider
 */
export async function addProvider(options: {
  id?: string;
  name?: string;
  displayName?: string;
  baseURL?: string;
  authType?: "api-key" | "bearer" | "custom";
  models?: string;
  quiet?: boolean;
}): Promise<void> {
  let providerId = options.id;
  let name = options.name;
  let displayName = options.displayName;
  let baseURL = options.baseURL;
  let authType = options.authType;
  let models = options.models;

  // Interactive mode if not quiet
  if (!options.quiet) {
    console.log();
    p.intro(pc.cyan("Add Custom Provider"));

    const existing = providerId ? await getUserProvider(providerId) : undefined;

    providerId = (await p.text({
      message: "Provider ID (unique identifier, e.g., 'openrouter', 'deepseek')",
      placeholder: existing?.id || "my-provider",
      initialValue: existing?.id,
      validate: (value) => {
        if (!value || value.trim().length === 0) {
          return "Provider ID cannot be empty";
        }
        if (!/^[a-z0-9-]+$/.test(value)) {
          return "Provider ID can only contain lowercase letters, numbers, and hyphens";
        }
      },
    })) as string;

    if (p.isCancel(providerId)) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }

    // Load existing if updating
    const existingProvider = await getUserProvider(providerId);

    name = (await p.text({
      message: "Provider name",
      placeholder: existingProvider?.name || "My Provider",
      initialValue: existingProvider?.name,
    })) as string;

    if (p.isCancel(name)) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }

    displayName = (await p.text({
      message: "Display name (shown in UI)",
      placeholder: existingProvider?.displayName || name,
      initialValue: existingProvider?.displayName || name,
    })) as string;

    if (p.isCancel(displayName)) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }

    baseURL = (await p.text({
      message: "Base URL (API endpoint)",
      placeholder: existingProvider?.baseURL || "https://api.example.com/v1",
      initialValue: existingProvider?.baseURL,
      validate: (value) => {
        if (!value || value.trim().length === 0) {
          return "Base URL cannot be empty";
        }
        try {
          new URL(value);
        } catch {
          return "Invalid URL format";
        }
      },
    })) as string;

    if (p.isCancel(baseURL)) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }

    authType = (await p.select({
      message: "Authentication type",
      options: [
        { value: "api-key", label: "API Key (X-API-Key header)" },
        { value: "bearer", label: "Bearer Token (Authorization: Bearer)" },
        { value: "custom", label: "Custom (no authentication or custom headers)" },
      ],
      initialValue: existingProvider?.authType || "bearer",
    })) as "api-key" | "bearer" | "custom";

    if (p.isCancel(authType)) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }

    models = (await p.text({
      message: "Default models (comma-separated, optional)",
      placeholder: existingProvider?.defaultModels?.join(", ") || "model-1, model-2",
      initialValue: existingProvider?.defaultModels?.join(", "),
    })) as string;

    if (p.isCancel(models)) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }
  }

  // Validate required fields
  if (!providerId || !name || !displayName || !baseURL || !authType) {
    showError(
      "Error: Missing required fields",
      "Usage: swixter providers add --id <id> --name <name> --display-name <display> --base-url <url> --auth-type <type>"
    );
  }

  const provider: ProviderPreset = {
    id: providerId,
    name: name,
    displayName: displayName,
    baseURL: baseURL,
    authType: authType,
    defaultModels: models ? models.split(",").map(m => m.trim()).filter(Boolean) : [],
    docs: "",
  };

  try {
    await upsertUserProvider(provider);

    if (!options.quiet) {
      p.outro(pc.green(`${MARKERS.success} Provider "${providerId}" added successfully!`));
    } else {
      showSuccess(`Provider "${providerId}" added successfully!`, {
        ID: providerId,
        "Display Name": displayName,
        "Base URL": baseURL,
        "Auth Type": authType,
      });
    }
  } catch (error) {
    showError(`Failed to add provider: ${error}`);
  }
}

/**
 * Remove a provider
 */
export async function removeProvider(providerId?: string, options?: { quiet?: boolean }): Promise<void> {
  if (!providerId) {
    if (options?.quiet) {
      showError("Error: Provider ID is required", "Usage: swixter providers remove <provider-id>");
    }

    // Interactive mode - show list and let user select
    console.log();
    p.intro(pc.cyan("Remove Custom Provider"));

    const userProviders = await loadUserProviders();

    if (userProviders.length === 0) {
      p.outro(pc.yellow("No user-defined providers to remove"));
      return;
    }

    providerId = (await p.select({
      message: "Select provider to remove",
      options: userProviders.map(p => ({
        value: p.id,
        label: `${p.displayName} (${p.id})`,
      })),
    })) as string;

    if (p.isCancel(providerId)) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }

    const confirmed = await p.confirm({
      message: `Are you sure you want to remove provider "${providerId}"?`,
    });

    if (p.isCancel(confirmed) || !confirmed) {
      p.cancel("Operation cancelled");
      process.exit(0);
    }
  }

  const deleted = await deleteUserProvider(providerId);

  if (!deleted) {
    showError(`Error: Provider "${providerId}" not found`, "Use 'swixter providers list' to see all providers");
  }

  if (!options?.quiet) {
    p.outro(pc.green(`${MARKERS.success} Provider "${providerId}" removed successfully!`));
  } else {
    showSuccess(`Provider "${providerId}" removed successfully!`);
  }
}

/**
 * Show provider details
 */
export async function showProvider(providerId?: string): Promise<void> {
  if (!providerId) {
    showError("Error: Provider ID is required", "Usage: swixter providers show <provider-id>");
  }

  const provider = await getUserProvider(providerId);

  if (!provider) {
    showError(`Error: Provider "${providerId}" not found`, "Use 'swixter providers list' to see all providers");
  }

  console.log();
  p.intro(pc.cyan(`Provider: ${provider.displayName}`));
  console.log();
  console.log(`  ID: ${pc.cyan(provider.id)}`);
  console.log(`  Name: ${pc.yellow(provider.name)}`);
  console.log(`  Display Name: ${pc.yellow(provider.displayName)}`);
  console.log(`  Base URL: ${pc.yellow(provider.baseURL)}`);
  console.log(`  Auth Type: ${pc.yellow(provider.authType)}`);
  if (provider.defaultModels && provider.defaultModels.length > 0) {
    console.log(`  Default Models:`);
    provider.defaultModels.forEach(model => console.log(`    - ${pc.dim(model)}`));
  }
  if (provider.docs) {
    console.log(`  Documentation: ${pc.blue(provider.docs)}`);
  }
  console.log();
}

File: packages/cli/crates/swixter/src/commands/providers.rs
use crate::cli::{ProviderAddArgs, ProvidersArgs, ProvidersCommand};
use crate::{EXIT_GENERAL, EXIT_INVALID_ARG, EXIT_NOT_FOUND, EXIT_SUCCESS};
use swixter_core::presets;
use swixter_core::types::{AuthType, ProviderPreset};
use swixter_core::user_providers;

pub fn dispatch(args: ProvidersArgs) -> i32 {
    match args.command {
        None | Some(ProvidersCommand::List) => list(),
        Some(ProvidersCommand::Add(a)) => add(a),
        Some(ProvidersCommand::Remove { id, quiet }) => remove(id, quiet),
        Some(ProvidersCommand::Show { id }) => show(&id),
    }
}

fn list() -> i32 {
    println!("Built-in providers:");
    for p in presets::builtin_presets() {
        println!("  {} — {}", p.id, p.display_name);
    }
    let user = user_providers::load();
    if !user.is_empty() {
        println!("User-defined providers:");
        for p in &user {
            println!("  {} — {}", p.id, p.display_name);
        }
    }
    EXIT_SUCCESS
}

fn add(a: ProviderAddArgs) -> i32 {
    if !a.quiet {
        eprintln!("Interactive provider add is not supported yet; use --quiet with flags.");
        return EXIT_INVALID_ARG;
    }
    let (id, name, display, base_url) = match (a.id, a.name, a.display_name, a.base_url) {
        (Some(i), Some(n), Some(d), Some(u)) => (i, n, d, u),
        _ => {
            eprintln!("✗ --id --name --display-name --base-url are required with --quiet");
            return EXIT_INVALID_ARG;
        }
    };
    if !id
        .chars()
        .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-')
        || id.is_empty()
    {
        eprintln!("✗ Invalid provider id ([a-z0-9-] only)");
        return EXIT_INVALID_ARG;
    }
    if presets::find_builtin(&id).is_some() {
        eprintln!("⚠ Overriding built-in provider \"{id}\"");
    }
    let auth_type = match a.auth_type.as_deref().unwrap_or("api-key") {
        "api-key" => AuthType::ApiKey,
        "bearer" => AuthType::Bearer,
        "custom" => AuthType::Custom,
        other => {
            eprintln!("✗ Invalid --auth-type: {other}");
            return EXIT_INVALID_ARG;
        }
    };
    let models = a
        .models
        .map(|m| {
            m.split(',')
                .map(|s| s.trim().to_string())
                .filter(|s| !s.is_empty())
                .collect()
        })
        .unwrap_or_default();
    let preset = ProviderPreset {
        id: id.clone(),
        name,
        display_name: display,
        base_url,
        default_models: models,
        auth_type,
        ..Default::default()
    };
    match user_providers::add(preset) {
        Ok(()) => {
            println!("✓ Provider \"{id}\" added");
            EXIT_SUCCESS
        }
        Err(e) => {
            eprintln!("✗ {e}");
            EXIT_GENERAL
        }
    }
}

fn remove(id: Option<String>, quiet: bool) -> i32 {
    let id = match id {
        Some(i) => i,
        None => {
            let user = user_providers::load();
            if user.is_empty() {
                println!("No user-defined providers.");
                return EXIT_SUCCESS;
            }
            let items: Vec<&str> = user.iter().map(|p| p.id.as_str()).collect();
            match dialoguer::Select::new()
                .with_prompt("Remove which provider?")
                .items(&items)
                .interact()
            {
                Ok(i) => user[i].id.clone(),
                // TS: 交互选择 cancel → exit 0
                Err(_) => return EXIT_SUCCESS,
            }
        }
    };
    if !quiet {
        let ok = dialoguer::Confirm::new()
            .with_prompt(format!("Remove provider \"{id}\"?"))
            .default(false)
            .interact()
            .unwrap_or(false);
        if !ok {
            return EXIT_SUCCESS;
        }
    }
    match user_providers::remove(&id) {
        Ok(true) => {
            println!("✓ Provider \"{id}\" removed");
            EXIT_SUCCESS
        }
        Ok(false) => {
            eprintln!("✗ Provider \"{id}\" not found");
            EXIT_NOT_FOUND
        }
        Err(e) => {
            eprintln!("✗ {e}");
            EXIT_GENERAL
        }
    }
}

fn show(id: &str) -> i32 {
    match presets::find_provider(id) {
        Some(p) => {
            println!("{} — {}", p.id, p.display_name);
            println!("  baseURL: {}", p.base_url);
            if let Some(c) = &p.base_url_chat {
                println!("  baseURLChat: {c}");
            }
            println!("  authType: {:?}", p.auth_type);
            if !p.default_models.is_empty() {
                println!("  models: {}", p.default_models.join(", "));
            }
            if let Some(k) = &p.env_key {
                println!("  env_key: {k}");
            }
            EXIT_SUCCESS
        }
        None => {
            eprintln!("✗ Provider \"{id}\" not found");
            EXIT_NOT_FOUND
        }
    }
}
